### PR TITLE
[rolling] Export YAML_CPP_DLL define on Windows (#30)

### LINKED ISSUE #10: [Windows] It is nice YAML_CPP_DLL is exported by default.
On Windows, the `yaml-cpp` is built as shared libraries. In such case, `YAML_CPP_DLL` is required to define to correctly consume the `yaml-cpp` (otherwise, [visibility decoration](https://github.com/jbeder/yaml-cpp/blob/4fb1c4b92bf8d94b32ebccdd890407d45b3bc794/include/yaml-cpp/dll.h#L19) won't be defined and it results in linkage time error).

And currently every project consuming `yaml-cpp` will need to define it in their CMake project to enable Windows build. It would be nice to see this exported directly from this package and all the downstream project can get the definition for free.

UPDATED:
For example,
* https://github.com/ros-perception/image_common/blob/ros2/camera_calibration_parsers/src/parse_yml.cpp#L49
* https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/src/rosbag2_transport/recorder.cpp#L36
* https://github.com/ros2/rviz/blob/ros2/rviz_common/src/rviz_common/yaml_config_reader.cpp#L42

And I discovered this issue when I try to [enable navigation2](https://github.com/ros-planning/navigation2/pull/1704) on Windows.

### PR DESCRIPTION
**Pull request by jacobperron**
Forward port #30 to Rolling.

This should fix #10

### CODE REVIEW

**Comment by jacobperron** [comment]
Requires the following changes to avoid build warnings:
- https://github.com/ros2/rosbag2/pull/964
- https://github.com/ros2/rviz/pull/831

**Comment by jacobperron** [comment]
CI with connected PRs:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17340/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11875/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17830/)

File: yaml_cpp_vendor-extras.cmake.in [hunk 1]
```diff
@@ -16,3 +16,6 @@ set(yaml_cpp_vendor_LIBRARIES ${YAML_CPP_LIBRARIES})
 set(yaml_cpp_vendor_INCLUDE_DIRS ${YAML_CPP_INCLUDE_DIR})
 
 list(APPEND yaml_cpp_vendor_TARGETS yaml-cpp)
+if(WIN32)
+  set_target_properties(${YAML_CPP_LIBRARIES} PROPERTIES INTERFACE_COMPILE_DEFINITIONS YAML_CPP_DLL)
+endif()
```
Comment (clalancette) [inline yaml_cpp_vendor-extras.cmake.in:21]: I think it is worthwhile to add a comment to explain a bit about what is going on here from our investigation.  Something like:

```suggestion
# On Windows, yaml-cpp 0.7.0 requires that downstream consumers set YAML_CPP_DLL and not yaml_cpp_EXPORT in order to set dllimport properly.
# This behavior will likely change in future versions of yaml-cpp.
if(WIN32)
  set_target_properties(${YAML_CPP_LIBRARIES} PROPERTIES INTERFACE_COMPILE_DEFINITIONS YAML_CPP_DLL)
endif()
```

**Review by clalancette — APPROVED** [review]
I've left a suggestion for a comment, just to capture some of the investigation we did yesterday.  I'm still approving since it is just a comment.  No need to rerun CI with this change; a clean Rpr job is enough.

File: yaml_cpp_vendor-extras.cmake.in [hunk 2]
```diff
@@ -16,3 +16,6 @@ set(yaml_cpp_vendor_LIBRARIES ${YAML_CPP_LIBRARIES})
 set(yaml_cpp_vendor_INCLUDE_DIRS ${YAML_CPP_INCLUDE_DIR})
 
 list(APPEND yaml_cpp_vendor_TARGETS yaml-cpp)
+if(WIN32)
+  set_target_properties(${YAML_CPP_LIBRARIES} PROPERTIES INTERFACE_COMPILE_DEFINITIONS YAML_CPP_DLL)
```
Comment (Ace314159) [inline yaml_cpp_vendor-extras.cmake.in:20]: I've identified the issue that was causing the build of `rosbag2_storage` to fail. It appears that `YAML_CPP_LIBRARIES` is now empty in this version of yaml-cpp. Using the target `yaml-cpp` directly fixes the issue.

```suggestion
  set_target_properties(yaml-cpp PROPERTIES INTERFACE_COMPILE_DEFINITIONS YAML_CPP_DLL)
```

File: yaml_cpp_vendor-extras.cmake.in (hunk 1 above)
Comment (jacobperron) [reply yaml_cpp_vendor-extras.cmake.in]: Added [1370b67](https://github.com/ros2/yaml_cpp_vendor/pull/38/commits/1370b679a18b186f3c72bae335561939e1983afc)

**Comment by jacobperron** [comment]
Windows: [Build Status](https://ci.ros2.org/job/ci_windows/)

**Pull request merged**